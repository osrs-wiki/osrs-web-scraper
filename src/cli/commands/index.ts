export { default as news } from "./news";
export { default as poll } from "./poll";
export { default as worlds } from "./worlds";
