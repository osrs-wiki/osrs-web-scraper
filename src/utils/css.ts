import { HTMLElement } from "node-html-parser";

/**
 * Extracts background image URL from CSS style attribute
 * Handles both single and double quotes, and HTML entity encoding
 */
export function extractBackgroundImageUrl(style: string): string | null {
  if (!style) return null;
  
  // First decode HTML entities
  const decodedStyle = style
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&amp;/g, '&');
  
  // Match background-image: url("...") or url('...') patterns
  const urlMatch = decodedStyle.match(/background-image\s*:\s*url\s*\(\s*['"']?([^'"')]+)['"']?\s*\)/i);
  
  if (urlMatch && urlMatch[1]) {
    return urlMatch[1];
  }
  
  return null;
}

/**
 * Extracts all background image URLs from style attributes in HTML elements
 */
export function extractBackgroundImages(elements: HTMLElement[]): string[] {
  const imageUrls: string[] = [];
  
  elements.forEach(element => {
    if (element.attributes?.style) {
      const url = extractBackgroundImageUrl(element.attributes.style);
      if (url) {
        imageUrls.push(url);
      }
    }
  });
  
  return imageUrls;
}