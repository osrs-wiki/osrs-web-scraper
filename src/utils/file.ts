import fs from "fs";
import client from "https";
import path from "path";

import { formatText } from "./text";

/**
 * Download a file from a url
 * @param url The url of the file to download
 * @param filepath The filepath of the downloaded file
 * @returns
 */
export const downloadFile = async (url: string, filepath: string) => {
  console.info(`Attempting file download: ${url}`);
  return new Promise((resolve, reject) => {
    client.get(url, (res) => {
      if (res.statusCode === 200) {
        res
          .pipe(fs.createWriteStream(filepath))
          .on("error", reject)
          .once("close", () => resolve(filepath));
        console.info(`Downloaded file ${filepath}`);
      } else {
        res.resume();
        reject(
          new Error(`Request Failed With a Status Code: ${res.statusCode}`)
        );
      }
    });
  });
};

/**
 * Get a file from a url
 * @param fileUrl The file url
 * @returns
 */
export const getFileName = (fileUrl: string) => {
  const parsed = new URL(fileUrl);
  return path.basename(parsed.pathname);
};

/**
 * Get the extension of a file from a url link.
 * @param fileUrl The file url
 * @returns
 */
export const getFileExtension = (fileUrl: string) => {
  return fileUrl.split(/[#?]/)[0].split(".").pop().trim();
};

const fileCharsReplaceMap: { [key: string]: string } = {
  ":": "-",
  "|": "-",
  "&nbsp;": "",
};

/**
 * Format a file name to ensure it is valid.
 * @param fileName The file name
 * @returns
 */
export const formatFileName = (fileName: string) => {
  const cleanedFileName = Object.keys(fileCharsReplaceMap).reduce(
    (name, find) => name.replaceAll(find, fileCharsReplaceMap[find]),
    fileName
  );
  return formatText(cleanedFileName);
};
