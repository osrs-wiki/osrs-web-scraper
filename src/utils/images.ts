import fs from "fs";
import client from "https";
import path from "path";

import { formatText } from "./text";

/**
 * Download an image from a url
 * @param url The url of the image to download
 * @param filepath The filepath of the downloaded image
 * @returns
 */
export const downloadImage = async (url: string, filepath: string) => {
  console.info(`Attempting image download: ${url}`);
  return new Promise((resolve, reject) => {
    client.get(url, (res) => {
      if (res.statusCode === 200) {
        res
          .pipe(fs.createWriteStream(filepath))
          .on("error", reject)
          .once("close", () => resolve(filepath));
        console.info(`Downloaded image ${filepath}`);
      } else {
        res.resume();
        reject(
          new Error(`Request Failed With a Status Code: ${res.statusCode}`)
        );
      }
    });
  });
};

/**
 * Get an image name from a url
 * @param imageUrl The image url
 * @returns
 */
export const getImageName = (imageUrl: string) => {
  const parsed = new URL(imageUrl);
  return path.basename(parsed.pathname);
};

/**
 * Get the extension of an image from a url link.
 * @param imageUrl The image url
 * @returns
 */
export const getImageExtension = (imageUrl: string) => {
  return imageUrl.split(/[#?]/)[0].split(".").pop().trim();
};

/**
 * Format a file name to ensure it is valid.
 * @param fileName The file name
 * @returns
 */
export const formatFileName = (fileName: string) => {
  return formatText(fileName.replaceAll(":", "-"));
};
