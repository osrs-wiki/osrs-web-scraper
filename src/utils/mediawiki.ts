import { MediaWikiBreak, MediaWikiContent } from "@osrs-wiki/mediawiki-builder";
/**
 * Trims the leading and trailing MediaWikiBreak elements from the contents.
 *
 * @param contents The contents to trim.
 * @returns The trimmed contents.
 */
export const trim = (contents: MediaWikiContent[]): MediaWikiContent[] => {
  if (!contents || contents.length === 0) {
    return contents;
  }
  while (
    contents[0] instanceof MediaWikiBreak ||
    contents[contents.length - 1] instanceof MediaWikiBreak
  ) {
    if (contents.length > 0 && contents[0] instanceof MediaWikiBreak) {
      contents.shift();
    }
    if (
      contents.length > 0 &&
      contents[contents.length - 1] instanceof MediaWikiBreak
    ) {
      contents.pop();
    }
  }
  return contents;
};

/**
 * Checks if the contents are empty.
 * @param contents The contents to check.
 * @returns True if the contents are empty, otherwise false.
 */
export const isEmpty = (
  contents: string | MediaWikiContent | MediaWikiContent[]
): boolean => {
  if (typeof contents === "string") {
    return contents.trim().length === 0;
  } else if (Array.isArray(contents)) {
    return contents.length === 0 || contents.every(isEmpty);
  } else if (contents instanceof MediaWikiContent) {
    return isEmpty(contents.children);
  }
  return false;
};

/**
 * Checks if the first child of the contents starts with the given string.
 *
 * @param contents The contents to check.
 * @param str The string to check against.
 * @returns True if the first child starts with the given string, otherwise false.
 */
export const startsWith = (
  contents: string | MediaWikiContent | MediaWikiContent[],
  str: string
): boolean => {
  if (isEmpty(contents)) {
    return false;
  }
  if (typeof contents === "string") {
    return contents.startsWith(str);
  }
  const childContents = Array.isArray(contents) ? contents[0] : contents;
  if (typeof childContents.children === "string") {
    return childContents.children.startsWith(str);
  }
  return startsWith(childContents.children, str);
};

/**
 * Recursively retrieves the first string content from the given MediaWikiContent.
 *
 * @param contents The MediaWikiContent to search through.
 * @returns The first string content found, or undefined if none is found.
 */
export const getFirstStringContent = (
  contents: MediaWikiContent
): MediaWikiContent | undefined => {
  if (typeof contents.children === "string") {
    return contents;
  } else if (contents.children instanceof MediaWikiContent) {
    return getFirstStringContent(contents.children);
  } else if (Array.isArray(contents.children) && contents.children.length > 0) {
    return getFirstStringContent(contents.children[0]);
  }
  return undefined;
};
