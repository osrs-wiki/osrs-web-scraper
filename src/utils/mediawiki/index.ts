export { default as MediaWikiBuilder } from "./builder";
export { default as MediaWikiContent } from "./content";

export * from "./contents";
