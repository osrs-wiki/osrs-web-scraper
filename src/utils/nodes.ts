import { Node } from "node-html-parser";

/**
 * Util function to prevent having to include ts-ignore every time rawTagName is referenced.
 * @param node
 * @returns
 */
export const getNodeTagName = (node: Node): string => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return node.rawTagName;
};
