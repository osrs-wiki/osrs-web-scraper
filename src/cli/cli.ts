import { Command } from "commander";

import { news, poll, worlds } from "./commands";
import packageJson from "../../package.json";

const cli = new Command();

cli.name("OSRS Web Scraper").description("").version(packageJson.version);

cli.addCommand(news);
cli.addCommand(poll);
cli.addCommand(worlds);

export default cli;
