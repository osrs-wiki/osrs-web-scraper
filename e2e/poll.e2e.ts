import * as fs from "fs";

import { runCLICommand, cleanupOutput } from "./utils/cli";

describe("Poll Command E2E Tests", () => {
  // Note: Tests assume the project is already built with `npm run build`

  afterEach(async () => {
    await cleanupOutput();
  });

  it("should show help for poll command", async () => {
    const result = await runCLICommand({
      command: "node",
      args: [
        "-r",
        "typescript-transform-paths/register",
        "./dist/src/index.js",
        "poll",
        "--help",
      ],
    });

    expect(result.exitCode).toBe(0);
    expect(result.stdout).toContain("Scrape an OSRS poll");
  });

  it("should require poll URL argument", async () => {
    const result = await runCLICommand({
      command: "node",
      args: [
        "-r",
        "typescript-transform-paths/register",
        "./dist/src/index.js",
        "poll",
      ],
    });

    expect(result.exitCode).toBe(1);
    expect(result.stderr).toContain(
      "error: missing required argument 'string'"
    );
  });

  // This test requires Chrome/Chromium and internet access
  const shouldSkipScraping =
    process.env.CI || process.env.PUPPETEER_SKIP_DOWNLOAD;

  (shouldSkipScraping ? it.skip : it)(
    "should scrape poll and generate MediaWiki content",
    async () => {
      const pollUrl =
        "https://secure.runescape.com/m=poll/a=13/oldschool/results?id=1708";

      const result = await runCLICommand({
        command: "node",
        args: [
          "-r",
          "typescript-transform-paths/register",
          "./dist/src/index.js",
          "poll",
          pollUrl,
        ],
        timeout: 180000, // 3 minutes for scraping
      });

      expect(result.exitCode).toBe(0);

      // Check that output files were created
      const outDir = "./out/poll";
      expect(fs.existsSync(outDir)).toBe(true);

      // Find the generated file
      const files = fs.readdirSync(outDir);
      expect(files.length).toBeGreaterThan(0);

      const pollFile = files.find((file: string) => file.endsWith(".txt"));
      expect(pollFile).toBeDefined();

      if (pollFile) {
        const filePath = `${outDir}/${pollFile}`;
        const content = fs.readFileSync(filePath, "utf-8");

        // Verify it contains MediaWiki poll content
        expect(content).toContain("{{PollWrapper|top}}");
        expect(content).toContain("{{PollWrapper|bottom}}");
        expect(content).toMatchSnapshot("poll-1708.txt");
      }
    }
  );
});
