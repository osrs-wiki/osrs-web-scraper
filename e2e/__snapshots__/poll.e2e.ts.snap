// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Poll Command E2E Tests should scrape poll and generate MediaWiki content: poll-1708.txt 1`] = `
"{{PollNotice|1708|8 August 2025|14 August 2025}}


{{PollWrapper|top}}
{{PollDescription|Previous skills have come with resources that support their unique gameplay. We've striven to make Sailing feel authentically Old School with this same kind of progression and new, thematic resources.

Many materials will be introduced as core parts of the new skill for which you already voted, but it's up to you how they should affect other parts of the game.

We think they'd add value to existing skills like Mining and Woodcutting, so we'd like to offer them as updates to those skills. However they could be obtained in other ways, such as purely via Sailing activities.

Some content may be delivered after the initial launch depending on development time, player feedback, and overall project scope.

This poll closes on Thursday 14th August.}}

{{PollTotal|113652}}
{{Poll
|Question = Which skillcape and icon should be added for Sailing?
|Answer1 = A
|Votes1 = 23851
|Answer2 = B
|Votes2 = 20227
|Answer3 = C
|Votes3 = 26105
|Answer4 = D
|Votes4 = 32826
|Answer5 = Skip question
|Votes5 = 10643
}}
{{Poll
|Question = Should the new trees be added to the game to let players chop and burn the new logs?
|Answer1 = Yes
|Votes1 = 100080
|Answer2 = No
|Votes2 = 10128
|Answer3 = Skip question
|Votes3 = 3444
}}
{{Poll
|Question = Should the new rocks be added to the game to let players mine the new ores?
|Answer1 = Yes
|Votes1 = 99755
|Answer2 = No
|Votes2 = 10504
|Answer3 = Skip question
|Votes3 = 3393
}}
{{Poll
|Question = Should Lantern Fishing be added to the game with Sailing?
|Answer1 = Yes
|Votes1 = 98962
|Answer2 = No
|Votes2 = 9680
|Answer3 = Skip question
|Votes3 = 5010
}}
{{Poll
|Question = Should Crab Hunting be introduced as a Hunter training method?
|Answer1 = Yes
|Votes1 = 100308
|Answer2 = No
|Votes2 = 9439
|Answer3 = Skip question
|Votes3 = 3905
}}
{{Poll
|Question = Should Coral Nurseries and new Hops Patch expansions be integrated into Farming?
|Answer1 = Yes
|Votes1 = 93310
|Answer2 = No
|Votes2 = 13123
|Answer3 = Skip question
|Votes3 = 7219
}}
{{Poll
|Question = Should new Crafting materials be added as described?
|Answer1 = Yes
|Votes1 = 98300
|Answer2 = No
|Votes2 = 9340
|Answer3 = Skip question
|Votes3 = 6012
}}
{{Poll
|Question = Should new blowpipes be added as described?
|Answer1 = Yes
|Votes1 = 89850
|Answer2 = No
|Votes2 = 16585
|Answer3 = Skip question
|Votes3 = 7217
}}
{{Poll
|Question = If the new blowpipes are added, should the Rosewood Blowpipe receive a Special Attack as described?
|Answer1 = Yes
|Votes1 = 85551
|Answer2 = No
|Votes2 = 17484
|Answer3 = Skip question
|Votes3 = 10617
}}
{{Poll
|Question = Should Anti-odour salt be added to the game?
|Answer1 = Yes
|Votes1 = 78528
|Answer2 = No
|Votes2 = 17907
|Answer3 = Skip question
|Votes3 = 17217
}}
{{Poll
|Question = Should Anti-bleed tonic be added to the game?
|Answer1 = Yes
|Votes1 = 81901
|Answer2 = No
|Votes2 = 17929
|Answer3 = Skip question
|Votes3 = 13822
}}
{{Poll
|Question = Should Super fishing potion be added to the game?
|Answer1 = Yes
|Votes1 = 90435
|Answer2 = No
|Votes2 = 17009
|Answer3 = Skip question
|Votes3 = 6208
}}
{{Poll
|Question = Should Extreme energy potion be added to the game?
|Answer1 = Yes
|Votes1 = 90611
|Answer2 = No
|Votes2 = 17585
|Answer3 = Skip question
|Votes3 = 5456
}}
{{Poll
|Question = Should Super hunter potion be added to the game?
|Answer1 = Yes
|Votes1 = 86614
|Answer2 = No
|Votes2 = 19433
|Answer3 = Skip question
|Votes3 = 7605
}}
{{Poll
|Question = Should Extended stamina potion be added to the game?
|Answer1 = Yes
|Votes1 = 96701
|Answer2 = No
|Votes2 = 12700
|Answer3 = Skip question
|Votes3 = 4251
}}
{{Poll
|Question = Should Armadyl brew be added to the game?
|Answer1 = Yes
|Votes1 = 88846
|Answer2 = No
|Votes2 = 13686
|Answer3 = Skip question
|Votes3 = 11120
}}
{{Poll
|Question = Should Red crab be added to the game?
|Answer1 = Yes
|Votes1 = 94628
|Answer2 = No
|Votes2 = 10711
|Answer3 = Skip question
|Votes3 = 8313
}}
{{Poll
|Question = Should Blue crab be added to the game?
|Answer1 = Yes
|Votes1 = 94651
|Answer2 = No
|Votes2 = 10705
|Answer3 = Skip question
|Votes3 = 8296
}}
{{Poll
|Question = Should Swordtip squid be added to the game?
|Answer1 = Yes
|Votes1 = 94315
|Answer2 = No
|Votes2 = 10756
|Answer3 = Skip question
|Votes3 = 8581
}}
{{Poll
|Question = Should Giant krill be added to the game?
|Answer1 = Yes
|Votes1 = 95235
|Answer2 = No
|Votes2 = 10315
|Answer3 = Skip question
|Votes3 = 8102
}}
{{Poll
|Question = Should Haddock be added to the game?
|Answer1 = Yes
|Votes1 = 93676
|Answer2 = No
|Votes2 = 10998
|Answer3 = Skip question
|Votes3 = 8978
}}
{{Poll
|Question = Should Rainbow crab be added to the game?
|Answer1 = Yes
|Votes1 = 93558
|Answer2 = No
|Votes2 = 12457
|Answer3 = Skip question
|Votes3 = 7637
}}
{{Poll
|Question = Should Yellowfin be added to the game?
|Answer1 = Yes
|Votes1 = 95153
|Answer2 = No
|Votes2 = 10617
|Answer3 = Skip question
|Votes3 = 7882
}}
{{Poll
|Question = Should Halibut be added to the game?
|Answer1 = Yes
|Votes1 = 94686
|Answer2 = No
|Votes2 = 11099
|Answer3 = Skip question
|Votes3 = 7867
}}
{{Poll
|Question = Should Bluefin be added to the game?
|Answer1 = Yes
|Votes1 = 95208
|Answer2 = No
|Votes2 = 10836
|Answer3 = Skip question
|Votes3 = 7608
}}
{{Poll
|Question = Should Marlin be added to the game?
|Answer1 = Yes
|Votes1 = 96244
|Answer2 = No
|Votes2 = 10239
|Answer3 = Skip question
|Votes3 = 7169
}}
{{Poll
|Question = Should new Smithing additions for Cannonballs be added to the game?
|Answer1 = Yes
|Votes1 = 99294
|Answer2 = No
|Votes2 = 10301
|Answer3 = Skip question
|Votes3 = 4057
}}
{{Poll
|Question = Should new planks for shipbuilding be added as described?
|Answer1 = Yes
|Votes1 = 94689
|Answer2 = No
|Votes2 = 13315
|Answer3 = Skip question
|Votes3 = 5648
}}
{{Poll
|Question = Should the new Thieving methods and related rewards be added to the game?
|Answer1 = Yes
|Votes1 = 102504
|Answer2 = No
|Votes2 = 7435
|Answer3 = Skip question
|Votes3 = 3713
}}

{{PollFooter}}
{{PollWrapper|bottom}}
"
`;
